Flask server for interacting with the contrib bots through outgoing webhooks.

vabs22 opened the pull request.

vabs22 commented:
Things to keep in mind:
1. The flask server needs to be run inside the vagrant machine.
2. It supports followup bot for now, the followup bot config details need to be added in the form specified in comments, in bot_config dict().
3. To make it work for other bots, the base_url in populate_db changes need to be updated to http://0.0.0.0:5002/bots/<botname> and the corresponding config of bot needs to be added into bot_config dict().

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -318,6 +318,14 @@ def handle(self, **options):
+            create_users(zulip_realm, zulip_outgoing_bots, bot_type=UserProfile.OUTGOING_WEBHOOK_BOT)
+            Service.objects.create(name="test",
+                                   user_profile=get_user_profile_by_email("outgoing-webhook@zulip.com"),
+                                   base_url="http://0.0.0.0:5002/bots/followup",
```
Why is this `0.0.0.0:5002`?  Seems like maybe we want it to be `127.0.0.1:5002` instead.

vabs22 replied:
okay

timabbott commented:
I merged a tweaked version of the outgoing webhook commit, so the basic outgoing webhook user should be available in the test suite now.

I think we should do a bit of cleanup of the flask server system next.  I think the main thing that this needs is a configuration system that we can use to move the API keys and other settings out of the flask server code and into a config file.  

I would suggest using a ConfigParser / zuliprc file style approach, similar to what our other bots do, though it might make sense to have a variation in the approach to support having multiple bots defined in the same file?

vabs22 commented on contrib_bots/contrib_server.py:
```diff
@@ -0,0 +1,75 @@
+    state_handler = StateHandler()
+
+    event = json.loads(request.data)
+    message_handler.handle_message(message=event["message"],
```
Yeah it would be helpful to add the comment. Currently we are sending the data directly but later on I think we would be passing it through an interface designed for this server. That might be the best place for it. Do you want me to add a TODO comment inside DoRestCall?

robot-dreams replied:
That sounds good for now.

vabs22 commented:
@timabbott @robot-dreams Thanks for you reviews!. I have incorporated the comments and updated it. 
I also think it would be reasonable to have a single file containing the configurations of all bots. I have completed an intial version for the same and added a new commit which generates a single config file for all active bots. The sample format of the config file is as above: 

![screenshot from 2017-05-30 23-16-46](https://cloud.githubusercontent.com/assets/7190228/26597216/e5f34d26-458e-11e7-8174-809bb176abeb.png)

For downloading the above file I have added a download button similar to existing one. It can be seen here: 

![screenshot from 2017-05-30 22-56-43](https://cloud.githubusercontent.com/assets/7190228/26597299/3970cbae-458f-11e7-94b8-7fed34f33193.png)

I have also updated the flask server to read the configuration file and then populate the `bots_config` dict using `SafeConfigParser`. Please share your thoughts.

robot-dreams commented on static/js/settings_bots.js:
```diff
@@ -56,6 +56,14 @@ exports.generate_zuliprc_content = function (email, api_key) {
 };
 
+exports.generate_zuliprc_content_for_single_file = function (email, api_key) {
+    return "[" + email.substring(0, email.indexOf("-bot@zulip.localhost")) + "]" +
```
Will this always be `@zulip.localhost`?  Would it make sense to use `email.indexOf("-bot@")` instead?

robot-dreams replied:
Also, would this code be more readable with a separate `bot_name_from_email` function?

vabs22 replied:
I agree with both of your points.

robot-dreams commented on static/templates/settings/bot-settings.handlebars:
```diff
@@ -59,6 +59,13 @@
         <div class="form-horizontal" id="api_key_button_box">
             <div class="input-group side-padded-container">
+                <div>
+                    <span>Download single zuliprc of all active bots</span>
```
Could the `zuliprc` name cause confusion because it's also used for other kinds of configuration files?

If so, would a name like `zulipbotrc` or `flaskbotrc` or (you can probably come up with something better) make sense?

robot-dreams replied:
Would it be helpful to say what this is for (e.g. `Download config of all active bots for Flask bot server` or something)?

vabs22 replied:
yeah `flaskbotrc` would make more sense to the users.

vabs22 replied:
yeah, I agree.

robot-dreams commented on contrib_bots/contrib_server.py:
```diff
@@ -0,0 +1,67 @@
+app = Flask(__name__)
+
+@app.route('/')
+def main():
```
Just checking, is it necessary to define this route / function?

If so, can it be called something besides `main()`?
If not, can it be removed?

robot-dreams commented on contrib_bots/contrib_server.py:
```diff
@@ -0,0 +1,67 @@
+
+from zulip import Client
+
+config_file_path = sys.argv[1]
```
Referencing `sys.argv[1]` at the top level like this seems like it could lead to confusion or problems—especially if this module gets imported somewhere else.

While this is pretty clearly a standalone app and probably won't ever be imported, I still think it's important to consistently use good habits across the codebase.

Instead of the current approach, do you think it would make sense to define a `load_config` function that takes a `file_path` as a parameter, and then call it from the `if __name__ == "__main__"` block?

vabs22 replied:
Yeah it would be much better to shift these code lines into separate functions and call them from `if __name__ == "__main__"` block

robot-dreams commented on contrib_bots/contrib_server.py:
```diff
@@ -0,0 +1,67 @@
+                    site=bots_config[bot]["site"])
+    restricted_client = BotHandlerApi(client)
+    message_handler = bots_lib_module[bot].handler_class()
+    state_handler = StateHandler()
```
Can you please add a `TODO` to handle stateful bots (e.g. the Tic Tac Toe bot)?

timabbott commented:
@vabs22 I merged the first commit to help make it easy for other folks to collaborate on the flask system.  I noticed a few things we'll want to fix in the near future:
* `python api/flask_bot_server.py` doesn't work; you have to run it directly from `api/`.  Since with the `pip` installer, this will be available from just about any directory, we definitely will want to work on this element.
* Structurally, `flask_api_server.py` might belong in the `api/zulip` directory; I'm not sure, and possibly also should be executable.  I didn't move it when merging because its path handling is fragile (see above) and I didn't want to debug potential issues there before making it possible for multiple folks to work on this code easily.
* We should use optparse/argparse to make `--help` work, so it's easy to figure it out.
* We might want to rename it (at least when installed) to have a reasonable name in PATH.  I.e. I think we kinda want the command after a `pip install` to be `zulip-bot-server ~/flaskbotrc`.  
* The error message if your `flaskbotrc` file has as invalid bot name is super confusing.  We should be really clear that the problem is that there is no bot with name "test" if your bot is called "test-bot", and that the most likely solution is to edit your `flaskbotrc` file to fix the name of the bot to match the bot you want to run.

@eeshangarg, as the maintainer of the `pip` package, can you collaborate with @vabs22 on these follow-up tasks (which I think we should try to get done this week, since they should mostly be quick.).  

More medium-term, we will want to work on these:
* Add a basic test suite for the Flask system so we can change it with minimal fear.
* supervisord configuration for running the flask bot persistently that's super easy to setup (e.g. `install-zulip-flask-supervisord <path_to_flaskbotrc>`
* We should work on writing trivial heroku instructions for using this.

@vabs22 can you and @robot-dreams open issues for all of the above follow-up tasks that aren't going to be done this week?  I imagine those will be your projects for the next week or two.

For the JS changes, I opened https://github.com/zulip/zulip/issues/5355 for a related cleanup we'll want to do, which would be reasonable for you to work on @vabs22.  The remaining patch in this PR I can merge once we fix a couple small things:
* "Download config of all active bots for Flask bot server" needs to be tagged for translation
* "Download config of all active bots for Flask bot server" is confusing.  I think we want to name the "Flask bot server" something more suggestive, e.g. "Zulip Botserver" (I don't think our users should care that's based on Flask).  Also, it should probably only download the outgoing webhook bots, since e.g. incoming webhooks aren't useful to put in this flask server.  So maybe something like "Download config of all active outgoing webhooks in Zulip Botserver format."

vabs22 commented:
I have updated this PR, the UI looks like: 
![screenshot from 2017-06-16 15-46-11](https://user-images.githubusercontent.com/7190228/27224048-cdc8899e-52b0-11e7-84df-b0279a6577d3.png)
Once the outgoing webhook UI gets merged, I will update the tests.

showell commented:
@vabs22 Can you clean up the commit message here?  The title line should be no longer than 65 characters, and then you can have an empty line and more text to elaborate as needed.

vabs22 commented:
@showell I have updated the commit message and the tests. Please take a look.

showell commented:
Merged!  Thanks, @vabs22!
